opendevin/condenser/condenser.py

```diff
@@ -0,0 +1,147 @@
+        logger.debug(
+            f'desired_token_count_to_summarize={desired_token_count_to_summarize}'
+        )
+        logger.debug(
+            f'len(candidate_messages_to_summarize)={len(candidate_messages_to_summarize)}'
+        )
+
+        if len(candidate_messages_to_summarize) == 0:
+            raise SummarizeError(
+                f"Summarize error: tried to run summarize, but couldn't find enough messages to compress [len={len(messages)}]"
+            )
+
+        # TODO: Try to make an assistant message come after the cutoff
+
+        message_sequence_to_summarize = candidate_messages_to_summarize
+
+        if len(message_sequence_to_summarize) <= 1:
+            # This prevents a potential infinite loop of summarizing the same message over and over
+            raise SummarizeError(
+                f"Summarize error: tried to run summarize, but couldn't find enough messages to compress [len={len(message_sequence_to_summarize)} <= 1]"
+            )
+        else:
+            print(
+                f'Attempting to summarize with last summarized event id = {last_summarized_event_id}'
+            )
+
+        summary_action: AgentSummarizeAction = self.summarize_messages(
+            message_sequence_to_summarize=message_sequence_to_summarize
+        )
+        summary_action.last_summarized_event_id = (
+            last_summarized_event_id if last_summarized_event_id else -1
+        )
+        return summary_action
+
+    def _format_summary_history(self, message_history: list[dict]) -> str:
+        # TODO use existing prompt formatters for this (eg ChatML)
+        return '\n'.join([f'{m["role"]}: {m["content"]}' for m in message_history])
+
+    def summarize_messages(self, message_sequence_to_summarize: list[Message]):
+        """Summarize a message sequence using LLM"""
+        context_window = self.config.max_input_tokens  # type: ignore
+        summary_prompt = SUMMARY_PROMPT_SYSTEM
+        summary_input = self._format_summary_history(
+            self.get_text_messages(message_sequence_to_summarize)  # type: ignore
+        )
+        summary_input_tkns = self.get_token_count(summary_input)  # type: ignore
+        if context_window is None:
+            raise ValueError('context_window should not be None')
+        if summary_input_tkns > MESSAGE_SUMMARY_WARNING_FRAC * context_window:
+            trunc_ratio = (
+                MESSAGE_SUMMARY_WARNING_FRAC * context_window / summary_input_tkns
+            ) * 0.8  # For good measure...
+            cutoff = int(len(message_sequence_to_summarize) * trunc_ratio)
+            summary_input = str(
+                [
+                    self.summarize_messages(
+                        message_sequence_to_summarize=message_sequence_to_summarize[
+                            :cutoff
+                        ]
+                    )
+                ]
+                + message_sequence_to_summarize[cutoff:]
```

Comment:
This doesn't seem to work well, we want to convert to strings all messages, but got:

`[AgentSummarizeAction(summarized_actions='I asked the user about the task, provided a concise description of the Moon, and wrote a Bash script to solve the Three Queens problem.', summarized_observations='The user requested a concise description of the Moon and a Bash script for the Three Queens problem. The script execution encountered issues, resulting in exit codes -1 and 1.', action='summarize'), <opendevin.llm.messages.Message object at 0x17a962030>, <opendevin.llm.messages.Message object at 0x17a961ac0>, <opendevin.llm.messages.Message object at 0x17a962330>, <opendevin.llm.messages.Message object at 0x17a962000>, <opendevin.llm.messages.Message object at 0x17a961b80>, <opendevin.llm.messages.Message object at 0x17a961f10>, <opendevin.llm.messages.Message object at 0x17a962270>, <opendevin.llm.messages.Message object at 0x17a961e20>, <opendevin.llm.messages.Message object at 0x17a9621b0>, <opendevin.llm.messages.Message object at 0x17a961e80>, <opendevin.llm.messages.Message object at 0x17a961be0>, <opendevin.llm.messages.Message object at 0x17a961c40>, <opendevin.llm.messages.Message object at 0x17a961ee0>, <opendevin.llm.messages.Message object at 0x17a961e50>, <opendevin.llm.messages.Message object at 0x17a9615b0>, <opendevin.llm.messages.Message object at 0x17a961d60>, <opendevin.llm.messages.Message object at 0x17a9611f0>, <opendevin.llm.messages.Message object at 0x17a961f40>, <opendevin.llm.messages.Message object at 0x17a962e40>]
`

I think it's when it's called a second time recursively.

Reply:
Yes, it happens in the second recursion. I made some corrections. 
When I am trying it out, I am not able to trigger this if statement. How are you doing it?😅

Reply:
I have a large context (11-12k) compared to the `max_input_tokens=4000` value. I suspect that's why it works for me, so you may want to play with the agent, keep adding to context, then switch to a much lower setting. This 12k is triple the 4k limit. (So the 0.75 factor in use in this PR, even if we subtract the initial prompt, is left waaaay behind. 😅)

If it helps, here is my event stream. It's not great for testing in general TBH, but I suspect it's good enough for this issue simply because of that size difference.
[eventstream_main.zip](https://github.com/user-attachments/files/16432010/eventstream_main.zip)

You may want to use all of [these settings](https://github.com/OpenDevin/OpenDevin/pull/2937#issuecomment-2257255702), like filestore (to save locally on disk the events in the stream, it will create a `main` sub-directory for it) and cli session enabled. Then replace the 'main' subdirectory in the filestore with the contents here.

I should note, I didn't use the UI. If you're using the UI, you can do the same, just need to set JWT token and use the sub-directory named with the right session-id, instead of "main" sub-directory. Frankly I've never tried to restore a cli session into a UI session, but it should work(tm). Most events are the same obviously, just the UI might have some extra INIT events, which shouldn't matter for restore.